ldes-server-compaction/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/compaction/application/services/PaginationCompactionService.java

```diff
@@ -0,0 +1,71 @@
+import be.vlaanderen.informatievlaanderen.ldes.server.compaction.domain.services.CompactableRelationPredicate;
+import be.vlaanderen.informatievlaanderen.ldes.server.compaction.domain.services.CompactionComparator;
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.model.LdesFragmentIdentifier;
+import be.vlaanderen.informatievlaanderen.ldes.server.fragmentation.entities.Fragment;
+import be.vlaanderen.informatievlaanderen.ldes.server.fragmentation.repository.FragmentRepository;
+import org.springframework.stereotype.Component;
+import be.vlaanderen.informatievlaanderen.ldes.server.domain.model.TreeRelation;
+import java.util.Optional;
+
+@Component
+public class PaginationCompactionService {
+
+	private final CompactableFragmentPredicate compactableFragmentPredicate;
+	private final CompactableRelationPredicate compactableRelationPredicate;
+	private final FragmentRepository fragmentRepository;
+	private final FragmentCompactionService fragmentCompactionService;
+
+	public PaginationCompactionService(FragmentRepository fragmentRepository,
+			FragmentCompactionService fragmentCompactionService) {
+		this.fragmentRepository = fragmentRepository;
+		this.fragmentCompactionService = fragmentCompactionService;
+		this.compactableFragmentPredicate = new CompactableFragmentPredicate();
+		this.compactableRelationPredicate = new CompactableRelationPredicate(this.fragmentRepository);
+
+	}
+
+	public void applyCompactionStartingFromNode(Fragment fragment) {
+		Fragment firstFragmentOfCompaction = fragment;
+		while (firstFragmentOfCompactionHasOutgoingRelations(firstFragmentOfCompaction)) {
+			Optional<Fragment> secondFragmentOfCompaction = retrieveMostCompactedRelationFragment(
+					firstFragmentOfCompaction);
+			if (secondFragmentOfCompaction.isPresent()) {
+				testAndApplyPossibleCompaction(firstFragmentOfCompaction, secondFragmentOfCompaction.get());
+				firstFragmentOfCompaction = secondFragmentOfCompaction.get();
+			} else {
+				break;
+			}
+		}
+	}
+
+	private void testAndApplyPossibleCompaction(Fragment firstFragmentOfCompaction,
+			Fragment secondFragmentOfCompaction) {
+		if (compactableRelationPredicate.test(firstFragmentOfCompaction)
+				&& compactableFragmentPredicate.test(secondFragmentOfCompaction)
+				&& compactableFragmentPredicate.test(firstFragmentOfCompaction)) {
+			fragmentCompactionService.compactFragments(firstFragmentOfCompaction, secondFragmentOfCompaction);
+		}
+	}
+
+	private boolean firstFragmentOfCompactionHasOutgoingRelations(Fragment firstFragmentOfCompaction) {
+		return !firstFragmentOfCompaction.getRelations().isEmpty();
+	}
+
+	private Optional<Fragment> retrieveMostCompactedRelationFragment(Fragment currentFragment) {
+		LdesFragmentIdentifier ldesFragmentIdentifier = currentFragment
+				.getRelations()
+				.stream()
+				.map(TreeRelation::treeNode)
+				.max(new CompactionComparator())
+				.orElse(currentFragment
```

Comment:
Isn't this 'orElse' only triggered when 'getRelations' is empty? Won't this result in an NPE when you do .get(0) below?

Reply:
We could solve this by remove the orElse and adding a map with 'fragmentRepo.retrieve...' and just returning this whole thing.

Reply:
fixed